### PR TITLE
Fix #10870: Chart.js new component

### LINKED ISSUE #5665: ChartsJS: TimePoint value type
## 1) Environment
- PrimeFaces version: 8.0

## 2) Expected behavior
Charts js allows to pass datapoints as x/y Tuples. Right now Primefaces implemented that using the NumericPoint class. Which works, but only accepts, oh wonder, numeric points ;-).
It would be greate if there would be also a DatePoint class, or something similar, which accepts, a combination of LocalDatetime and Numeric points. 
e.g.
x = 2020-01-01-12-00-00
y = 123.45

Charts Js accepts that without any problems and shows them properly when using the type "time" for an axes.

## 3) Actual behavior
I'ts not possible to pass a Date (or LocalDatetime or ...) value to chars js. I have to figure out the conversation on my own...

### LINKED ISSUE #10849: Charts: unified and consistent data interface classes
### Description

As a developer of charts that often stack multiple elements (bar, line) to show the relationship between 2 or more data sets where my number of data points is not consistently equal; I want a consistent way to feed data to the Cartesian chart types that preserve my x / y tuples so that I do not need to pre parse all my data points to get a distinct list of X axis labels, then build each data set ensuring that I provide a value for each X value I found across all datasets, but instead let chart.js figure it out for me, instead of needing to write a potentially error prone hack to massage all the data points to fit the primefaces wrapper when the underlying chart.js can handle the tuples.

### Describe the solution you would like

For each of the ChartDataSet implementations, have the data field accept a List<Object> instead of List<Number> (or a list of an abstraction that makes sense to collect x / y tuples)  like the LineChartDataSet currently does so that if a developer wants to pass in a list of NumericPoint, or a class of their own that has a toString implementation that produces JSON that will output correctly when processed by ChartUtils.writeDataValue() the same way BubblePoint and NumericPoint do.

### Additional context

The specific classes that would need to be modified to need a minimally viable fix are:  BarChartDataSet, PolarAreaChartDataSet, and RadarChartDataSet.

### LINKED ISSUE #10870: ChartJS: new approach
### Description

Based on discussions #8819 and #10849

I'll try to sum up the big lines:
- Have one and only component: `<p:chart />`. There is no need to have multiple components, at worse `<p:chart />` should redirect to the proper renderer depending the type: lucky for us chartJS does the job for us.
- Model abstraction: As of today, PF is responsible to implement chartJS API into Java. First, chartJS API is freaking big! Second, It's hard to follow chartjs model changes from one version to another, which makes us left behind compare to the version we are already using: this is frustrating for users.
- KISS:  `<p:chart />` could be a very simple JSF wrapper of chartJS, take a String as input, and handle a few ajax events out of the box. As it takes a String, users are free to either implement their own API and serialize it into JSON, or use existing solutions like https://github.com/xdev-software/chartjs-java-model (which is already 4.4 compatible and java 11 compatible since 1.0.2) 
- Deprecate current impl for 14 

### Describe the solution you would like

**XHTML**

```xml
<div class="card">
    <h5>Plain JSON</h5>
    <p:chart value="#{chartJsView.barModel}" style="width: 100%; height: 500px;">
        <p:ajax event="itemSelect" listener="#{chartJsView.itemSelect}" update="growl"/>
    </p:chart>
</div>
<div class="card">
    <h5>XDEV ChartJS model</h5>
    <p:chart value="#{chartJsView.barModel2}" style="width: 100%; height: 500px;">
        <p:ajax event="itemSelect" listener="#{chartJsView.itemSelect}" update="growl"/>
    </p:chart>
</div>
```

**JAVA**
```java
    public void createBarModel() {
        barModel = "{" +
            "    \"type\": \"bar\"," +
            "    \"data\": {" +
            "        \"datasets\": [" +
            "            {" +
            "                \"type\": \"bar\"," +
            "                \"data\": [" +
            "                    65," +
            "                    59," +
            "                    80," +
            "                    81," +
            "                    56," +
            "                    55," +
            "                    40" +
            "                ]," +
            "                \"label\": \"My First Dataset\"," +
            "                \"hidden\": false," +
            "                \"backgroundColor\": \"rgba(255, 99, 132, 0.2)\"," +
            "                \"borderColor\": \"rgb(255, 99, 132)\"," +
            "                \"borderWidth\": 1" +
            "            }," +
            "            {" +
            "                \"type\": \"bar\"," +
            "                \"data\": [" +
            "                    85," +
            "                    69," +
            "                    20," +
            "                    51," +
            "                    76," +
            "                    75," +
            "                    10" +
            "                ]," +
            "                \"label\": \"My Second Dataset\"," +
            "                \"hidden\": false," +
            "                \"backgroundColor\": \"rgba(255, 159, 64, 0.2)\"," +
            "                \"borderColor\": \"rgb(255, 159, 64)\"," +
            "                \"borderWidth\": 1" +
            "            }" +
            "        ]," +
            "        \"labels\": [" +
            "            \"January\"," +
            "            \"February\"," +
            "            \"March\"," +
            "            \"April\"," +
            "            \"May\"," +
            "            \"June\"," +
            "            \"July\"" +
            "        ]" +
            "    }," +
            "    \"options\": {" +
            "        \"responsive\": true," +
            "        \"maintainAspectRatio\": false," +
            "        \"barPercentage\": 0.9," +
            "        \"indexAxis\": \"x\"," +
            "        \"scales\": {" +
            "            \"y\": {" +
            "                \"offset\": true," +
            "                \"stacked\": false," +
            "                \"reverse\": false," +
            "                \"beginAtZero\": true," +
            "                \"ticks\": {" +
            "                    \"autoSkip\": true," +
            "                    \"mirror\": false" +
            "                }" +
            "            }" +
            "        }," +
            "        \"plugins\": {" +
            "            \"title\": {" +
            "                \"display\": true," +
            "                \"text\": \"Bar Chart\"" +
            "            }" +
            "        }" +
            "    }" +
            "}";
    }

   // using XDEV ChartJS model
    public void createBarModel2() {
        barModel2 = new BarChart()
            .setData(new BarData()
                .addDataset(new BarDataset()
                    .setData(65, 59, 80, 81, 56, 55, 40)
                    .setLabel("My First Dataset")
                    .setBackgroundColor(new Color(255, 99, 132, 0.2))
                    .setBorderColor(new Color(255, 99, 132))
                    .setBorderWidth(1))
                .addDataset(new BarDataset()
                    .setData(85, 69, 20, 51, 76, 75, 10)
                    .setLabel("My Second Dataset")
                    .setBackgroundColor(new Color(255, 159, 64, 0.2))
                    .setBorderColor(new Color(255, 159, 64))
                    .setBorderWidth(1)
                )
                .setLabels("January", "February", "March", "April", "May", "June", "July"))
            .setOptions(new BarOptions()
                .setResponsive(true)
                .setMaintainAspectRatio(false)
                .setIndexAxis(BarOptions.IndexAxis.X)
                .setScales(new Scales().addScale(Scales.ScaleAxis.Y, new BarScale < CategoryTicks > ()
                    .setBarPercentage(BigDecimal.valueOf(0.9))
                    .setStacked(false)
                    .setTicks(new CategoryTicks()
                        .setAutoSkip(true)
                        .setMirror(true))))
                .setPlugins(new Plugins()
                    .setTitle(new software.xdev.chartjs.model.options.Title()
                        .setDisplay(true)
                        .setText("Bar Chart using XDEV java model")))
            ).toJson();
    }
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by melloware**
Fix #10870: Chart.js new component
Fix #5665 ChartJS Time Values
Fix #10849 Unify Charts

- [x] Documentation Created
- [x] All new Showcase examples showing Raw JSON and XDEV Model
- [x] Recreated all showcase examples using XDEV
- [x] Custom example so you can paste any JSON in you want
- [x] Facet use `<f:facet name="value>` for simpler charts
- [x] Add zoom plugin https://www.chartjs.org/chartjs-plugin-zoom/latest/guide/
- [x] Compress charts.js

### CODE REVIEW

**Comment by melloware** [comment]
I created a Custom example in the Showcase where you can take any Chart.js example code and paste it in and it renders the chart.

image

File: primefaces/pom.xml [hunk 1]
```diff
@@ -607,6 +606,7 @@
                         <!-- uncompressed widgets -->
                         <resourcesSet>
                             <includes>
+                                <include>chart/**</include>
```
Comment (Rapster) [inline primefaces/pom.xml:610]: does it get minified then?

File: docs/14_0_0/components/chart.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+    public String getJson() {
+        return json;
+    }
+}
+```
+
+### Getting Started with XDEV Java Model
+
+You can use [XDEV Chart.js Java Model](https://github.com/xdev-software/chartjs-java-model) to render a server side model of your Chart.
```
Comment (Rapster) [inline docs/14_0_0/components/chart.md:206]: should we indicate the XDEV version compatible with PF?

File: docs/14_0_0/components/chart.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+```
+
+```java
+public class Bean {
+    
+    private String json;
+
+    @PostConstruct
+    public void init() {
+        json = "{\r\n"
```
Comment (Rapster) [inline docs/14_0_0/components/chart.md]: can't we use a text block for this (e.g """) ?

File: primefaces/pom.xml (hunk 1 above)
Comment (melloware) [reply primefaces/pom.xml:610]: No I moved this to the unminified section so it gets combined but not minified.

File: docs/14_0_0/components/chart.md (hunk 2 above)
Comment (melloware) [reply docs/14_0_0/components/chart.md:206]: Let me add it to the Dependencies section of the docs where we have OpenLibre and other optional libraries.

File: docs/14_0_0/components/chart.md (hunk 3 above)
Comment (melloware) [reply docs/14_0_0/components/chart.md]: I would love to but I think that is Java 17.

File: docs/14_0_0/gettingstarted/whatsnew.md [hunk 4]
```diff
@@ -13,6 +13,9 @@ Look into [migration guide](https://primefaces.github.io/primefaces/14_0_0/#/../
     * OS settings for `prefers-reduced-motion: reduce` is now respected and PF disables all animations.
     * Minimum inline Ajax load animation duration via `PrimeFaces.ajax.minLoadAnimation`.
 
+* Chart
+   * A new more flexible version of Chart.js can be used `<p:chart>` allowing raw JSON or XDEV Java Model
+
```
Comment (Rapster) [inline docs/14_0_0/gettingstarted/whatsnew.md:18]: should we mention that the current impl is deprecated and will be remov in later version?

File: docs/14_0_0/components/chart.md (hunk 3 above)
Comment (Rapster) [reply docs/14_0_0/components/chart.md]: ah my bad, I had in mind it was java 11...

File: docs/14_0_0/gettingstarted/whatsnew.md (hunk 4 above)
Comment (Rapster) [inline docs/14_0_0/gettingstarted/whatsnew.md]: should we mention that the current impl is deprecated and will be remov in later version?

Comment (melloware) [reply docs/14_0_0/gettingstarted/whatsnew.md]: Yes let me add a migration guide warning

**Comment by melloware** [comment]
I also added the `<f:facet name="value">` per @tandraschko suggestion and I like it for simple charts where you don't even need Java code.

File: primefaces/pom.xml (hunk 1 above)
Comment (Rapster) [reply primefaces/pom.xml:610]: why not minify it?

Comment (melloware) [reply primefaces/pom.xml:610]: So the 2 Chart JS libs the CHartJS itself and the moment plugin are already minified.   If you try and minify it again the Resource Minifier throws an error about the level of Javascript.   So the only thing not being minified is the Widget itself its just being appended on so its small.   That is how the current CHarts.js stuff is happening too because of this issue it was always unminified just combined to 1 file

File: primefaces/src/main/java/org/primefaces/component/chart/ChartRenderer.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            facet.encodeAll(context);
+
+            // restore the original writer
+            context.setResponseWriter(originalWriter);
+            return fsw.toString();
+        }
+        else {
+            return chart.getValue();
+        }
+    }
```
Comment (Rapster) [inline primefaces/src/main/java/org/primefaces/component/chart/ChartRenderer.java:109]: you had no issues with escaping? I remember doing something very similar but had escaping issues...

Comment (melloware) [reply primefaces/src/main/java/org/primefaces/component/chart/ChartRenderer.java:109]: Not that I have seen so far. There may be an issue but I have not run into it yet.

File: primefaces/pom.xml (hunk 1 above)
Comment (Rapster) [reply primefaces/pom.xml:610]: I assume you're refering to this: https://github.com/primefaces/primefaces/pull/10871#issuecomment-1774091993 and my PR contained the workaround. Ideally everything should be minified (whether it's small or big),

Comment (melloware) [reply primefaces/pom.xml:610]: Yes so I saw what you did and it does not work.  You tried to compress `2-chartjs-widget.js` and then later combine them into `charts.js` with the 2 minified files and `2-widget.js` but that won't work.  It Just combines them but the widget is still uncompressed.

**Comment by astappiev** [comment]
Could it be possible to include [chartjs-plugin-zoom](https://www.chartjs.org/chartjs-plugin-zoom/latest/) as well?

File: primefaces/pom.xml (hunk 1 above)
Comment (Rapster) [reply primefaces/pom.xml:610]: [chart.js.txt](https://github.com/primefaces/primefaces/files/13975438/chart.js.txt)

I can see 3 files in here and everything look compressed, am I missing something?

**Comment by melloware** [comment]
I don't see why not @astappiev as multiple people have asked for this before.

File: primefaces/pom.xml (hunk 1 above)
Comment (melloware) [reply primefaces/pom.xml:610]: Let me try again but that is not the behavior I am seeing inside the final JAR.  Maybe I didn't do a `clean`

**Comment by melloware** [comment]
@Rapster compression is working fine thanks!

@astappiev added Zoom plugin

**Comment by Rapster** [comment]
Great job! 💯

**Comment by melloware** [comment]
Final size compressed is 241 KB!

**Comment by melloware** [comment]
@Rapster @astappiev i was also able to accomplish this now!

1. Old and new Charts can co-exist now. I removed the old chart.js file and now they all share the same CHart.js.  All of the demos are working properly and now charts can co-exist!

2. I marked the old ones deprecated.

image

File: docs/14_0_0/components/chart.md (hunk 3 above)
Comment (Rapster) [reply docs/14_0_0/components/chart.md]: For the sake of readibility, and as it is only documentation, here using text blocks would be better

File: docs/14_0_0/gettingstarted/dependencies.md [hunk 6]
```diff
@@ -18,6 +18,7 @@ any 3rd part work incorporated are compatible with the PrimeFaces Licenses.
 | barcode4j-light | org.primefaces.extensions | barcode4j-light | 2.3.0 | Optional | Barcode |
 | qrcodegen | io.nayuki | qrcodegen | 1.8.0 | Optional | QR Code support for Barcode |
 | owasp-java-html-sanitizer | com.googlecode.owasp-java-html-sanitizer | owasp-java-html-sanitizer |  20220608.1 | Optional | TextEditor |
+| XDev chartjs-java-model | software.xdev | chartjs-java-model |  1.1.2 | Optional | Chart |
```
Comment (Rapster) [inline docs/14_0_0/gettingstarted/dependencies.md:21]: `XDev chartjs-java-model`  is not even an optional dependency, don't think it should be mention here

File: docs/14_0_0/components/chart.md (hunk 3 above)
Comment (melloware) [reply docs/14_0_0/components/chart.md]: Good idea

**Comment by melloware** [comment]
@Rapster you can submit any follow PR's if you find I have missed anything.

**Review by Rapster — COMMENTED** [review]
forgot to submit the review...

File: docs/14_0_0/gettingstarted/dependencies.md (hunk 6 above)
Comment (melloware) [reply docs/14_0_0/gettingstarted/dependencies.md:21]: I will remove
